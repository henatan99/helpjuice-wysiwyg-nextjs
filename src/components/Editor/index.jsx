import { useEffect, useState, useRef } from "react"
import { h1Style, pStyle } from "../../utils/block_styles"

const Editor = () => {
    const initialStyle = {
        width: 'fit-content',
        fontSize: '20px',
        border: 'solid 1px black',
        height: 'fit-content',
        outline: 'none',
        padding: '1rem',
        display: 'inline-block',
        tabIndex: 0
    }

    const initialBlocks = {
        lastBlockId: 1,
        list: [
            {id: 1, content: "Hi there", style: {initialStyle}},
        ]
    } 

    const blocksRef = useRef(null);

    useEffect(() => {
        blocksRef.current.children[0].focus();
    }, []);
  
    const [ editable, setEditable ] = useState(true)
    const [ blocks, setBlocks ] = useState(initialBlocks)
    const [ currentBlock, setCurrentBlock ] = useState(blocks.list[0])

    console.log(blocks)

    const setFocus = (activeIndex) => {
        blocksRef.current.children[0].focus()
    }

    const handleChange = (e) => {
        const char = e.target.innerText[e.target.innerText.length - 1]
        console.log('blocks', blocks);
        console.log('currentBlock', currentBlock)
        console.log('boolean', 1 == currentBlock.id)
        const blockIndex = blocks.list.indexOf(block => block.id == 1) 
        console.log('my block index', blockIndex)
        if(char == 0) {
            let updatedBlocksList = blocks.list
            
            setBlocks(
                {
                    ...blocks,
                    list: [
                        ...updatedBlocksList, 
                        
                    ],
                    lastBlockId: blocks.lastBlockId + 1,
                }
            )
            setCurrentBlock({
                id: blocks.lastBlockId + 1, 
                content: "", 
                style: {...initialStyle, ...pStyle}, 
                focus: true,
                tabIndex: 1
            })

            setFocus(blocks.list.indexOf(block => block.id == blockIndex + 1))
        } 
        else if(blocks[blockIndex] === '/p') {
            setCurrentBlock(blocks.list[blocks.list.lastBlockId + 1])
            setBlocks(
                {
                    ...blocks,
                    list: [
                        ...blocks.list, 
                        {
                            id: blocks.lastBlockId + 1, 
                            content: "", 
                            style: {...initialStyle, ...pStyle}, 
                            focus: true,
                            tabIndex: 1
                        }
                    ],
                    lastBlockId: blocks.lastBlockId + 1,
                }
            )
        }
        else if(blocks[blockIndex] === '/1') {
            setCurrentBlock(blocks.list[blocks.list.lastBlockId + 1])
            setBlocks(
                {
                    ...blocks,
                    list: [
                        ...blocks.list, 
                        {
                            id: blocks.lastBlockId + 1, 
                            content: "", 
                            style: {...initialStyle, ...pStyle}, 
                            focus: true,
                            tabIndex: 1
                        }
                    ],
                    lastBlockId: blocks.lastBlockId + 1,
                }
            )
        }

        else {
            let updatedBlocksList = [...blocks.list]

            console.log('block index', currentBlock.id)
            updatedBlocksList[blockIndex] = { 
                ...updatedBlocksList[blockIndex], 
                content: updatedBlocksList[blockIndex] && updatedBlocksList[blockIndex].content ? 
                    updatedBlocksList[blockIndex].content + char : char
            }  
            setCurrentBlock(updatedBlocksList)
            setBlocks(
                {
                    ...blocks,
                    list: updatedBlocksList
                }
            )
        }
    }

    return (
        <div ref={blocksRef} >
            <button onClick={() => setEditable(!editable)}>Lock</button>
            {
                blocks.list && blocks.list.map(block => {
                    const { style } = block
                    return (
                        <EditableDiv 
                            editable={editable}
                            style={style}
                            handleChange={handleChange}
                        />
                    )
                })
            }
            <div>{JSON.stringify(currentBlock)}</div>
        </div>
    )
}

const EditableDiv = (props) => {
    const { editable, style, handleChange } = props
    
    return (
        <>
            <div 
                style={style} 
                contentEditable={editable} 
                placeholder="|" 
                onInput={handleChange}
            >
                {}
            </div>
        </>
    )
}

export default Editor 