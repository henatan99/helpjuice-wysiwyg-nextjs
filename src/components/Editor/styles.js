import styled from "styled-components";

export const StyledContainer = styled.div`
  width: 100%;
  position: relative;
`

export const StyledInput = (({currentStyle}) => {
    styled.input`
    ...currentStyle,
    padding: 0,
    border-width: 0,
    width: '101%',
    max-width: 'unset',
    color: '#636a76',
    margin: '-1%',
    padding-left: '1%',
    outline: 'none'
`
}) 

export const StyledModalWrapper = styled.div`
    marginTop: 0.5rem
`
