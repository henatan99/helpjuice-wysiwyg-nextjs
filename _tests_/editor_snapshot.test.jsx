import { render } from '@testing-library/react'
import Editor from '../src/components/Editor'

it('renders editor unchanged', () => {
  const { container } = render(<Editor />)
  expect(container).toMatchSnapshot()
})