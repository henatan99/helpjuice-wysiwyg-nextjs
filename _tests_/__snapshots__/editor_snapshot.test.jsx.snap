// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders editor unchanged 1`] = `
<div>
  <div
    class="sc-bczRLJ cHvIj"
  >
    <div
      style="font-size: 1rem; margin-bottom: 1rem; line-height: 1.8rem; font-family: Arial, Helvetica, sans-serif; position: relative;"
    />
    <textarea
      placeholder="Type / for blocks, @ to link docs or people"
      rows="20"
      style="padding: 0px 0px 0px 1%; border-width: 0px; width: 101%; max-width: unset; color: rgb(99, 106, 118); margin: -1%; outline: none; font-family: inherit;"
    />
  </div>
</div>
`;
